### PR TITLE
More Clustering Support

### PR DESCRIPTION
**Pull request by DominiqueMakowski**
This is a WIP overhaul of the clustering support of easystats. I'll list the features and changes here:

### Main function: n_clusters()

- `n_clusters()` has been fixed and improved

``` r
library(parameters)

n <- n_clusters(iris[, 1:4], package = c("easystats", "NbClust", "mclust"))
n
#> # Method Agreement Procedure:
#> 
#> The choice of 2 clusters is supported by 13 (43.33%) methods out of 30 (Elbow, Silhouette, Gap_Maechler2012, CH, Cindex, DB, Beale, Ratkowsky, PtBiserial, McClain, Dunn, SDindex, Mixture).
summary(n)
#>   n_Clusters n_Methods
#> 1          0         2
#> 2          2        13
#> 3          3         6
#> 4          4         1
#> 5          5         3
#> 6          7         2
#> 7         13         1
#> 8         15         2
head(as.data.frame(n))
#>   n_Clusters           Method   Package
#> 1          0           Hubert   NbClust
#> 2          0           Dindex   NbClust
#> 3          2            Elbow easystats
#> 4          2       Silhouette easystats
#> 5          2 Gap_Maechler2012 easystats
#> 6          2               CH   NbClust
plot(n)
```

<sup>Created on 2021-08-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### Specific methods

Specific methods that are commonly used were added for convenience and all. Some of them (elbow, gap, silhouette) are included in the above `n_clusters()`. The plots could certainly be improved but for a first draft it does the job.

``` r
library(parameters)

# Elbow method -----------------------------
x <- n_clusters_elbow(iris[1:4])
x
#> The Elbow method, that aims at minimizing the total intra-cluster variation (i.e., the total within-cluster sum of square), suggests that the optimal number of clusters is 2.
as.data.frame(x)
#>    n_Clusters       WSS
#> 1           1 596.00000
#> 2           2 220.87929
#> 3           3 138.88836
#> 4           4 113.64981
#> 5           5  90.22782
#> 6           6  95.25396
#> 7           7  72.75296
#> 8           8  64.61603
#> 9           9  59.48502
#> 10         10  59.10865
#> 11         11  59.38389
#> 12         12  68.74352
#> 13         13  35.91391
#> 14         14  41.47813
#> 15         15  33.59736
plot(x)
```

``` r
#
# Gap method ------------------------------
x <- n_clusters_gap(iris[1:4])
x
#> The Gap method, that compares the total intracluster variation of k clusters with their expected values under null reference distribution of the data, suggests that the optimal number of clusters is 2.
as.data.frame(x)
#>    n_Clusters       Gap         SE
#> 1           1 0.2214401 0.02507538
#> 2           2 0.4740829 0.03319823
#> 3           3 0.4942913 0.02579692
#> 4           4 0.4409764 0.02755421
#> 5           5 0.4017140 0.02235386
#> 6           6 0.4498149 0.02273822
#> 7           7 0.4483496 0.02347120
#> 8           8 0.4623633 0.02329825
#> 9           9 0.4767078 0.02639123
#> 10         10 0.4714515 0.02166452
#> 11         11 0.4203812 0.02390283
#> 12         12 0.3406179 0.02219453
#> 13         13 0.5486026 0.02587561
#> 14         14 0.4818785 0.02482117
#> 15         15 0.5270261 0.02403941
plot(x)
```

``` r
#
# Silhouette method --------------------------
x <- n_clusters_silhouette(iris[1:4])
x
#> The Silhouette method, based on the average quality of clustering, suggests that the optimal number of clusters is 2.
as.data.frame(x)
#>    n_Clusters Silhouette
#> 1           1  0.0000000
#> 2           2  0.5817500
#> 3           3  0.4599482
#> 4           4  0.4188923
#> 5           5  0.3536197
#> 6           6  0.3445645
#> 7           7  0.3152814
#> 8           8  0.3462564
#> 9           9  0.3180890
#> 10         10  0.3187029
#> 11         11  0.2794598
#> 12         12  0.3985803
#> 13         13  0.3590287
#> 14         14  0.2922473
#> 15         15  0.3347490
plot(x)
```

``` r
#
# DBSCAN method -------------------------
# NOTE: This actually primarily estimates the 'eps' parameter, the number of
# clusters is a side effect (it's the number of clusters corresponding to
# this 'optimal' EPS parameter).
x <- n_clusters_dbscan(iris[1:4], method = "kNN", min_size = 0.05) # 5 percent
x
#> The DBSCAN method, based on the total clusters sum of quares, suggests that the optimal eps = 1.56155019367095 (with min. cluster size set to 8), which corresponds to 1 clusters.
head(as.data.frame(x))
#>   n_Obs       eps
#> 1     1 0.2960431
#> 2     2 0.3100027
#> 3     3 0.3118788
#> 4     4 0.3118788
#> 5     5 0.3379073
#> 6     6 0.3518639
plot(x)
```

``` r
x <- n_clusters_dbscan(iris[1:4], method = "SS", eps_n = 100, eps_range = c(0.1, 2))
x
#> The DBSCAN method, based on the total clusters sum of quares, suggests that the optimal eps = 1.40505050505051 (with min. cluster size set to 15), which corresponds to 2 clusters.
head(as.data.frame(x))
#>         eps n_Clusters total_SS
#> 1 0.1000000          0      596
#> 2 0.1191919          0      596
#> 3 0.1383838          0      596
#> 4 0.1575758          0      596
#> 5 0.1767677          0      596
#> 6 0.1959596          0      596
plot(x)
```

``` r
#
# hclust method -------------------------
# iterations should be higher for real analyses
x <- n_clusters_hclust(iris[1:4], iterations = 50, ci = 0.90)
x
#> The bootstrap analysis of hierachical clustering highlighted 34 significant clusters.
head(as.data.frame(x))
#>   Cluster Row
#> 1       1  11
#> 2       1  49
#> 3       2  81
#> 4       2  82
#> 5       3   5
#> 6       3  38
plot(x)
```

<sup>Created on 2021-08-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

## K-Means

- [x] **model_parameters**
- [x] **model_performance**
- [x] **cluster_analysis** (this is a home-made wrapper to conveniently and consistently get "parameters", performance and cluster predictions)
- [x] Plotting

``` r
model <- kmeans(iris[1:4], centers = 3)

parameters::parameters(model)
#> # K-means Cluster Means
#> 
#> The 3 clusters accounted for 79.05% of the total variance of the original data.
#> 
#> Cluster | n_Obs | Sum_Squares | Sepal.Length | Sepal.Width | Petal.Length | Petal.Width
#> ---------------------------------------------------------------------------------------
#> 1       |    21 |       17.67 |         4.74 |        2.90 |         1.79 |        0.35
#> 2       |    33 |        6.43 |         5.18 |        3.62 |         1.47 |        0.27
#> 3       |    96 |      118.65 |         6.31 |        2.90 |         4.97 |        1.70

performance::performance(model)
#> # Indices of model performance
#> 
#> Sum_Squares_Within | Sum_Squares_Between | Sum_Squares_Total | Iterations
#> -------------------------------------------------------------------------
#> 681.371            |             142.754 |           538.617 |      2.000

parameters::cluster_analysis(iris[1:4], n=3, method = "kmeans")
#> # K-means Cluster Means
#> 
#> The 3 clusters accounted for 76.70% of the total variance of the original data.
#> 
#> Cluster | n_Obs | Sum_Squares | Sepal.Length | Sepal.Width | Petal.Length | Petal.Width
#> ---------------------------------------------------------------------------------------
#> 1       |    50 |       47.35 |        -1.01 |        0.85 |        -1.30 |       -1.25
#> 2       |    53 |       44.09 |        -0.05 |       -0.88 |         0.35 |        0.28
#> 3       |    47 |       47.45 |         1.13 |        0.09 |         0.99 |        1.01
#> 
#> # Indices of model performance
#> 
#> Sum_Squares_Within | Sum_Squares_Between | Sum_Squares_Total | Iterations
#> -------------------------------------------------------------------------
#> 596.000            |             138.888 |           457.112 |      3.000
#> 
#> # You can access the predicted clusters via 'predict()'.
```

<sup>Created on 2021-08-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### CODE REVIEW

**Comment by mattansb** [comment]
Shouldn't the plot elbow be on 3? I never understood how to make sense of those plots...

**Comment by DominiqueMakowski** [comment]
I suppose it depends on the method used to numerically find the "elbow". Here I used the minimum of the gradient:

https://github.com/easystats/parameters/blob/84a3709594841238d1cbb70a994a4d9412e80014/R/n_clusters_easystats.R#L16-L17

But possibly it should be the this + 1...? I don't know and it's quite hard to find some good explanations I must say

**Comment by mattansb** [comment]
No idea.... Maybe it's different for factors vs clusters?

**Comment by DominiqueMakowski** [comment]
## Support for Hierarchical clustering (hclust and pvclust)

``` r
library(parameters)

# Hierarchical clustering (hclust) ---------------------------
data <- iris[1:4]
model <- hclust(dist(data))
clusters <- cutree(model, 3)

rez <- model_parameters(model, data, clusters)
rez
#> # K-means Cluster Means
#> 
#> The 3 clusters accounted for 86.86% of the total variance of the original data.
#> 
#> Cluster | n_Obs | Sum_Squares | Sepal.Length | Sepal.Width | Petal.Length | Petal.Width
#> ---------------------------------------------------------------------------------------
#> 1       |    50 |       15.15 |         5.01 |        3.43 |         1.46 |        0.25
#> 2       |    72 |       64.62 |         6.55 |        2.96 |         5.27 |        1.85
#> 3       |    28 |        9.75 |         5.53 |        2.64 |         3.96 |        1.23

# pvclust (finds "significant" clusters) ---------------------------
data <- iris[1:4]
# NOTE: pvclust works on transposed data
model <- pvclust::pvclust(datawizard::data_transpose(data),
                          method.dist="euclidean",
                          nboot = 50,
                          quiet = TRUE)

rez <- model_parameters(model, data, ci = 0.90)
rez
#> # K-means Cluster Means
#> 
#> The 2 clusters accounted for 77.26% of the total variance of the original data.
#> 
#> Cluster | n_Obs | Sum_Squares | Sepal.Length | Sepal.Width | Petal.Length | Petal.Width
#> ---------------------------------------------------------------------------------------
#> 1       |    50 |       15.15 |         5.01 |        3.43 |         1.46 |        0.25
#> 2       |   100 |      139.80 |         6.26 |        2.87 |         4.91 |        1.68
```

<sup>Created on 2021-08-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>


## cluster_analysis: plotting and hclust

The previous method was baked into the `cluster_analysis` wrapper. One of the arguments that justifies the existence of this wrapper is that the models alone do not withhold the necessary information to do much postprocessing (performance, plotting, ...), whereas `cluster_analysis()` gives access to everything (in particular, the raw data).

Of note here is that I added a little (but exported) `cluster_performance()`, here in parameters (😱). Why you might ask? "it should be in performance". Indeed, but essentially, `cluster_analysis` also returns indices of performance, as most of these indices emerge from the parameters. More than that, performance for clusters usually needs the parameters to be computed (it's just some basic stats from the parameters). So my initial solution was to call `performance::model_performance` but then we ended up in a weird scenario in which model_performance called parameters() but cluster_analysis called performance(), and we had inter-dependencies which, as we know from experience, leads to hell. So by having cluster_performance in parameters, alongside the other clustering functions/helpers, we don't need to call `performance`'s functions. And it facilitates the support in performance, in we can simply do `model_performance.kmeans <- parameters::cluster_performance`. TLDR; the performance computation is now done in parameters rather than performance to avoid cross-dependencies. 

``` r
# Cluster analysis --------------------------------------------------------
library(parameters)

# K-Means ====================================================
rez <- cluster_analysis(iris[, 1:4], n = 3, method = "kmeans")
rez  # Show results
#> # K-means Cluster Means
#> 
#> The 3 clusters accounted for 76.70% of the total variance of the original data.
#> 
#> Cluster | n_Obs | Sum_Squares | Sepal.Length | Sepal.Width | Petal.Length | Petal.Width
#> ---------------------------------------------------------------------------------------
#> 1       |    47 |       47.45 |         1.13 |        0.09 |         0.99 |        1.01
#> 2       |    53 |       44.09 |        -0.05 |       -0.88 |         0.35 |        0.28
#> 3       |    50 |       47.35 |        -1.01 |        0.85 |        -1.30 |       -1.25
#> 
#> # Indices of model performance
#> 
#> Sum_Squares_Total | Sum_Squares_Between | Sum_Squares_Within |    R2
#> --------------------------------------------------------------------
#> 596.000           |             457.112 |            138.888 | 0.767
#> 
#> # You can access the predicted clusters via 'predict()'.
predict(rez)  # Get clusters
#>   [1] 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3
#>  [38] 3 3 3 3 3 3 3 3 3 3 3 3 3 1 1 1 2 2 2 1 2 2 2 2 2 2 2 2 1 2 2 2 2 1 2 2 2
#>  [75] 2 1 1 1 2 2 2 2 2 2 2 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 1 2 1 1 1 1 2 1 1 1 1
#> [112] 1 1 2 2 1 1 1 1 2 1 2 1 2 1 1 2 1 1 1 1 1 1 2 2 1 1 1 2 1 1 1 2 1 1 1 2 1
#> [149] 1 2
plot(rez)
```



``` r

# Hierarchical Clustering (hclust) ===========================
rez <- cluster_analysis(iris[, 1:4], n = 3, method = "hclust")
rez  # Show results
#> # K-means Cluster Means
#> 
#> The 3 clusters accounted for 74.35% of the total variance of the original data.
#> 
#> Cluster | n_Obs | Sum_Squares | Sepal.Length | Sepal.Width | Petal.Length | Petal.Width
#> ---------------------------------------------------------------------------------------
#> 1       |    49 |       40.12 |        -1.00 |        0.90 |        -1.30 |       -1.25
#> 2       |    24 |       18.65 |        -0.40 |       -1.36 |         0.06 |       -0.04
#> 3       |    77 |       94.08 |         0.76 |       -0.15 |         0.81 |        0.81
#> 
#> # Indices of model performance
#> 
#> Sum_Squares_Total | Sum_Squares_Between | Sum_Squares_Within |    R2
#> --------------------------------------------------------------------
#> 596.000           |             443.143 |            152.857 | 0.744
#> 
#> # You can access the predicted clusters via 'predict()'.
predict(rez)  # Get clusters
#>   [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
#>  [38] 1 1 1 1 2 1 1 1 1 1 1 1 1 3 3 3 2 3 2 3 2 3 2 2 3 2 3 3 3 3 2 2 2 3 3 3 3
#>  [75] 3 3 3 3 3 2 2 2 2 3 3 3 3 2 3 2 2 3 2 2 2 3 3 3 2 2 3 3 3 3 3 3 2 3 3 3 3
#> [112] 3 3 3 3 3 3 3 3 2 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3
#> [149] 3 3
plot(rez)
```



``` r
# n = NULL uses pvclust() to identify significant clusters =======
rez <- cluster_analysis(iris[, 1:4], n = NULL, method = "hclust")
plot(rez)
```



<sup>Created on 2021-08-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>


## Issues

Questions for @bwiernik 😁 

- Regarding the numerical method to determine the "elbow" (https://github.com/easystats/parameters/pull/578#issuecomment-895003506), is there any "correct" way over another?
- When I standardize the data, the explained variance changes, even if the clusters stay the same. My intuition is that it should remain the same. It could be an issue in the implementation (it could have to do with the centering of the variables). I based my implementation on [this code](https://stackoverflow.com/questions/68714612/compute-between-clusters-sum-of-squares-bcss-and-total-sum-of-squares-manually/68719858#68719858), currently implemented here:

https://github.com/easystats/parameters/blob/f7c98dc38b0108dcb44879676d7b817f2b4e202a/R/cluster_performance.R#L69-L76


Example:

``` r
k <- kmeans(iris[1:4], 3)
clusters <- k$cluster

# Raw
centers <- parameters::cluster_centers(iris[1:4], clusters)
attributes(centers)$variance
#> [1] 0.7904906

# Z
centers <- parameters::cluster_centers(datawizard::standardize(iris[1:4]), clusters)
attributes(centers)$variance
#> [1] 0.6796959
```

<sup>Created on 2021-08-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>


- Outliers: unsupervised methods like hclust or dbscan can identify "outliers", i.e., points not belonging to any clusters (grouped under the so-called "cluster 0"). When there is such pseudo-cluster, I'm not sure how to adjust the computations of the sum of squares, and thus of the variance explained...

**Comment by DominiqueMakowski** [comment]
We now support DBSCAN:

``` r
library(parameters)

rez <- cluster_analysis(iris[1:4], method = "dbscan", dbscan_eps = 1.45)
rez  # Show results
#> # K-means Cluster Means
#> 
#> The 3 clusters accounted for 58.75% of the total variance of the original data.
#> 
#> Cluster | n_Obs | Sum_Squares | Sepal.Length | Sepal.Width | Petal.Length | Petal.Width
#> ---------------------------------------------------------------------------------------
#> 0       |     8 |       77.55 |         0.75 |        0.41 |         0.31 |        0.13
#> 1       |    47 |       30.99 |        -1.03 |        0.82 |        -1.30 |       -1.26
#> 2       |    95 |      137.29 |         0.45 |       -0.44 |         0.62 |        0.61
#> 
#> # Indices of model performance
#> 
#> Sum_Squares_Total | Sum_Squares_Between | Sum_Squares_Within |    R2
#> --------------------------------------------------------------------
#> 596.000           |             350.173 |            168.280 | 0.588
#> 
#> # You can access the predicted clusters via 'predict()'.
predict(rez)  # Get clusters
#>   [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1
#>  [38] 1 1 1 1 0 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 0 2 2 2 2 2 2 2 2 2 2 2 2 2
#>  [75] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
#> [112] 2 2 2 2 2 2 0 0 2 2 2 0 2 2 2 2 2 2 2 2 0 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
#> [149] 2 2
plot(rez)
```



<sup>Created on 2021-08-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

- [x] I also added a vignette on clustering.
- [x] The clustering plots now deals correctly with "outliers"

**Comment by mattansb** [comment]
> Why you might ask? "it should be in performance"

It sounds like maybe the `model_parameters` itself should actually be `model_performance`....?

We should be thinking about the users' experience; they don't care that technically all the "performance" indices are computed in {parameters}, they will expect a performance function to be in performance.

**Comment by DominiqueMakowski** [comment]
> they will expect a performance function to be in performance.

Yup, wherever possible yes... but in this case, having the "performance" in performance would lead to recursive logic and direct cross-dependency (i.e., model_parameters would call model_performance which calls model_parameters 💥 ). RE user experience, if they look in performance (once this PR is merged and all), they will see that `model_performance.kmeans <- parameters::cluster_performance()` so they will know where to look for its content. We can also mention that in the docs!

**Comment by mattansb** [comment]
If the functionality is also exported to {performance} that's better, I guess.

But maybe better would be to have:

```R
pars <- parameters::model_parameters(kmeans)
performance::model_performance.parameters_clusters(pars )
```

Or something?

I mean, we've moved all the `check_*` functions to {performance} even when they didn't really have anything to do with evaluating fit or model adequacy. Consistency is key ^_^

**Comment by IndrajeetPatil** [comment]
> we've moved all the check_* functions to {performance} even when they didn't really have anything to do with evaluating fit or model adequacy

A few still remain, and we need to decide if we want to make this breaking change (moving them to `{performance}`) for the next release itself:

image

**Comment by DominiqueMakowski** [comment]
I suggest we see the reorganization after that PR is merged (ready soon, but the checks should pass first (but they are broken on main so 🤷)) and we see clearer how to best reorganize

File: R/cluster_centers.R [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+#' @param clusters A vector with clusters assignments (must be same length as rows in data).
+#' @param fun What function to use, \code{mean} by default.
+#' @param ... Other arguments to be passed to or from other functions.
+#'
+#'
+#' @examples
+#' k <- kmeans(iris[1:4], 3)
+#' cluster_centers(iris[1:4], clusters = k$cluster)
+#' cluster_centers(iris[1:4], clusters = k$cluster, fun = median)
+#' @importFrom stats aggregate
```
Comment (strengejacke) [inline R/cluster_centers.R:16]: please remove `@importFrom` tag

File: R/cluster_centers.R [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  # Get n obs
+  params <- data.frame(table(clusters))
+  names(params) <- c("Cluster", "n_Obs")
+
+  # Get Within clusters sum of squares (WCSS)
+  ss <- .cluster_analysis_SS(data, clusters)
+  params$Sum_Squares <- ss$WSS
+
+  # Get Cluster Centers
+  centers <- aggregate(data, list(Cluster=clusters), fun)
```
Comment (strengejacke) [inline R/cluster_centers.R:29]: add `stats::` here to `aggregate()`.

File: R/cluster_centers.R [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  attr(params, "variance") <- ss$BSS / ss$TSS
+
+  params
+}
+
+
+
+# Performance -------------------------------------------------------------
+
+#' @importFrom stats dist
```
Comment (strengejacke) [inline R/cluster_centers.R:43]: please remo `@importFrom` tag here

**Review by strengejacke — CHANGES_REQUESTED** [review]

**Review by strengejacke — APPROVED** [review]

**Comment by strengejacke** [comment]
What's the state of this PR, ready to merge? Does it require updates from any other packages?

**Comment by DominiqueMakowski** [comment]
It can be merged, it's a fairly self-contained set of features. Only requires the latest datawizard (which needs to be submitted first @IndrajeetPatil since correlation and modelbased also depend on it)

**Comment by strengejacke** [comment]
Can't resolve the conflicts here...

**Pull request merged**